enhancements/upstream-downstream-split.md

```diff
@@ -0,0 +1,228 @@
+
+### Downstreaming
+1. Change pushed to upstream/x/master
+2. Nightly job syncs upstream master with downstream using an [`UPSTREAM_MERGE.sh`](https://github.com/jmrodri/scripts/blob/master/UPSTREAM-MERGE.sh#L69) script for each subtree
+``` bash
+# UPSTREAM_MERGE.sh snippet
+if ! git diff-index HEAD --exit-code --quiet 2>&1 ; then
+	die "Working tree has modifications.  Cannot add."
+fi
+if ! git diff-index --cached HEAD --exit-code --quiet 2>&1 ; then
+	die "Index has modifications.  Cannot add."
+fi
+
+ref="${3:-master}"
+
+# check state of working directory
+git diff-index --quiet HEAD || { printf "!! Git status not clean, aborting !!\\n\\n%s" "$(git status)"; exit 1; }
+
+# update remote
+git fetch -t "$upstream_remote" "$ref"
+
+rev=$(git rev-parse $default --revs-only FETCH_HEAD) || exit $?
+
+if test "${#rev}" -ne 1; then
+	die "Multiple revisions found at HEAD: Expecting one, Got: '$rev'"
+fi
+
+git merge --no-commit -Xsubtree=$subtree_dir $rev
+
+# unmerged files are overwritten with the upstream copy
+unmerged_files=$(git diff --name-only --diff-filter=U --exit-code)
+differences=$?
+if [[ $differences -eq 1 ]]; then
+  unmerged_files_oneline=$(echo "$unmerged_files" | paste -s -d ' ')
+  git checkout --theirs -- $unmerged_files_oneline
+  if [[ $(git diff --check) ]]; then
+    echo "All conflict markers should have been taken care of, aborting."
+    exit 1
+  fi
+  git add -- $unmerged_files_oneline
+else
+  unmerged_files="<NONE>"
+fi
+
+echo "$rev" > "$subtree_dir/UPSTREAM_VERSION"
+git add UPSTREAM-VERSION
+
+# just to make sure an old version merge is not being made
+git diff --staged --quiet && { echo "No changed files in merge?! Aborting."; exit 1; }
+
+git commit -m "Merge upstream ref $ref" -m "Merge executed via ./UPSTREAM-MERGE.sh $upstream_remote $ref" -m "$(printf "Overwritten conflicts:\\n%s" "$unmerged_files")"
+
+```
+3. Carry over patches from downstream/master to new release branch.
+
+### Upstreaming
+Ideally, all changes should be made upstream. However, in case of a critical feature or fix for the downstream is blocked by the state of upstream, the change may be added as a patch to downstream first. A list of these patches will be maintained in a `patches` directory in the downstream monorepo. All patches in this directory are considered patches that need to be upstreamed.
+
+If these critical changes are not pushed upstream in some cohesive way, there is a high risk this will result in higher set difference between upstream and downstream over time. Prior to creating a new release, this patches directory will be compared with the state of upstream to remove the ones that have already been upstreamed. The remaining will be cherry-picked into the latest downstream release.
+
+While creating the downstream build, the vendor directory will be updated after the upstream sync with `go mod vendor`, after which the patches will be applied. This will preserve any downstream vendor changes.
```

Comment:
If we apply the patches after `go mod vendor` I think they may be ineffective -- I'm fairly sure a build will use whatever is in the `vendor` directory at the time. Am I perhaps misunderstanding something here?